Add post comparing closures with JavaScript

jonhoo opened the pull request:
Takes care of one of the points in #6.

This got a bit longer than I anticipated, but there is *a lot* to cover on the topic of closures. I think I've struck the right balance between useful and detailed, but feedback is welcome.

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+`fn` modifies its argument before returning it, then we will still be
+modifying the original list. Furthermore, our `map` doesn't check the
+type of its arguments at all — who knows what would happen if `list`
+wasn't an array, or `fn` wasn't a function…
```
Can we expand upon this? Maybe there's a way in JS to check that they are the correct types and throw an error otherwise? Like using typeof?

mgattozzi commented on JavaScript/closures.md:
````diff
@@ -0,0 +1,191 @@
+}
+```
+
+If we run this code we (again, unsurprisingly) get the same result as
````
Just have it say "we get the same result as our JavaScript version"

mgattozzi commented on JavaScript/closures.md:
````diff
@@ -0,0 +1,191 @@
+Now, for our `map` function:
+
+```rust
+fn map<T1, T2, F: Fn(&T1) -> T2>(list: &[T1], f: F) -> Vec<T2> {
````
I'd say also return a slice here but I'm not sure that's possible because of the borrow checker. Why not a `Vec` as input out of curiosity?

jonhoo replied:
We can't return a slice, because the slice would need to point to something we construct *inside* the function, which isn't possible. We do not take a `Vec` input simply because we don't need to; a slice is the most general type we could accept here (maybe except `Iterator`), so that's what I used. Even if we did take a `Vec`, we couldn't re-use that `Vec` as our output, both because the types wouldn't work out (`Vec<T1> != Vec<T2>`), and because that would entail modifying the input, which `map` doesn't normally do.

mgattozzi replied:
Ah I see. Using the primitive slice would be better in that case. That makes sense

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+we're converting to. This is `T1` and `T2` respectively.
+
+Second, we can't just directly place `Fn(&T1) -> T2` as the type of the
+`f` argument. Why not? Well, it's a little bit complicated, but The Book
```
Can we bold The Book and have it say **The Rust Book** instead for clarity?

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+Essentially, the compiler needs to know which function to *actually*
+call in order to generate the code for `map`. It's not satisfied by just
+anything that can act as a function. We could change our code to instead
+take a function *pointer* (see The Book linked to above), but that's a
```
Could we just link directly to it in the book and have pointer be the link?

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+topic for another time. The code as written above will actually cause
+the compiler to generate a dedicated, and *specialized* version of `map`
+**for every function it is ever called with**. This can vastly improve
+performance, though it also increases the size of your binary.
```
I didn't know that. That's pretty neat actually!

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+Closures in Rust are a fairly involved topic, and there is *a lot* more
+to be said about them. The examples above should be enough to get you
+off the ground though. I suggest you read the chapter on Closures in
+[The Book](https://doc.rust-lang.org/book/closures.html), and in
```
The Rust Book again

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+
+Closures in Rust are a fairly involved topic, and there is *a lot* more
+to be said about them. The examples above should be enough to get you
+off the ground though. I suggest you read the chapter on Closures in
```
How about "You should read" instead so that the article doesn't reference anyone in particular, especially since this might be edited by others in the future.

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+### Similarities between the two
+
+ - Both languages have anonymous functions (yay!), and functions are
+   first-class primitives, meaning we can write higher-order functions
```
I understand what first-class primitives are and you mentioned higher-order functions earlier with the link, but someone else might not be. Explaining it might help or at least linking an article. Not everyone coming to Rust (especially from JS where for many it's their first language) will know the terms.

mgattozzi commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+   `FnOnce` traits](https://doc.rust-lang.org/std/ops/trait.FnMut.html)).
+ - The compiler can, in many cases, generate more efficient code, since
+   the code will be specialized for the exact types involved. In
+   JavaScript, there is no way to know a priori what kind of functions
```
"to know beforehand" We want articles to have a conversational tone and less academic, even if we do bring things up like higher-order functions.

mgattozzi commented:
Also side note and unrelated to the PR I saw you are at MIT and Cambridge. I'm actually in Boston myself so if you're at the meetup tomorrow come say hi!

jonhoo commented on JavaScript/closures.md:
```diff
@@ -0,0 +1,191 @@
+the use of *callbacks*. Whenever an operation takes some time to
+complete, or you need to wait for an external event like the user
+pressing a key or a network connection to be made, you pass in an
+unnamed (anonymous) function to be executed when execution can continue.
```
"execution can continue" is more appropriate, since it may also *not* be waiting for an external event. I'll change "to be executed" to "to be called" though, which will help

mgattozzi replied:
Yes that wording is better!

jonhoo commented:
I believe 12ca8a5 addresses all your comments.

I think we actually met briefly during a past meetup in Boston, but I may be misremembering?
I wasn't aware the was a meetup tomorrow! When and where?

jonhoo commented:
Ah, found it! https://www.meetup.com/BostonRust/events/236669336/

mgattozzi commented:
Yup that's the one! I think we might have briefly. Either way definitely stop by if you can make it

mgattozzi commented:
This looks good to me so I'm going to merge it!
